apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.ts

```diff
@@ -0,0 +1,688 @@
+    new_time.append(
+      '<input class="form-control time-limit-field" type="number" name="time_add" style="width: 5em" value="5">',
+    );
+    const time_ref_select = $('<select class="form-control time-limit-field" name="time_ref">');
+    time_ref_select.append('<option value="minutes">minutes</option>');
+    if (row.time_remaining_sec !== null) {
+      time_ref_select.append('<option value="percent">% total limit</option>');
+    }
+    new_time.append(time_ref_select);
+    form.append(new_time);
+    if (row.has_closed_instance) {
+      const checkbox = $(
+        '<div class="form-check mb-2 reopen-closed-field"><input class="form-check-input" type="checkbox" name="reopen_closed" value="true" id="reopen-closed"><label class="form-check-label" for="reopen-closed">Also re-open closed instances</label></div>',
+      );
+      checkbox.toggle(row.time_remaining_sec === null);
+      form.append(checkbox);
+    }
+    const buttons = $('<div class="btn-toolbar pull-right">');
+    const cancel_button = $('<button type="button" class="btn btn-secondary mr-2">Cancel</button>');
+    cancel_button.on('click', function () {
+      $(that).popover('hide');
+    });
+    buttons.append(cancel_button);
+    buttons.append('<button type="submit" class="btn btn-success">Set</button>');
+    form.append(buttons);
+    form.on('submit', (e) => {
+      window.popoverSubmitViaAjax(e, that as any);
+    });
+    return form;
+  }
+
+  function scorebarFormatter(score: number) {
+    if (score != null) {
+      let bar = '<div class="progress bg" style="min-width: 5em; max-width: 20em;">';
+      let left = '',
+        right = '';
+      if (score >= 50) {
+        left = `${Math.floor(score)}%`;
+      } else {
+        right = `${Math.floor(score)}%`;
+      }
+      bar += `<div class="progress-bar bg-success" style="width: ${Math.floor(Math.min(100, score))}%">${left}</div>`;
+      bar += `<div class="progress-bar bg-danger" style="width: ${100 - Math.floor(Math.min(100, score))}%">${right}</div>`;
+      bar += '</div>';
+      return bar;
+    } else {
+      return '';
+    }
+  }
+
+  function listFormatter(list: string[]) {
+    if (!list || !list[0]) list = ['(empty)'];
+    return '<small>' + list.join(', ') + '</small>';
+  }
+
+  function uniqueListFormatter(list: string[]) {
+    if (!list || !list[0]) list = ['(empty)'];
+    const uniq = Array.from(new Set(list));
+    return '<small>' + uniq.join(', ') + '</small>';
+  }
+
+  function timeRemainingLimitFormatter(value: string, row: AssessmentInstanceRow) {
+    const container = $('<span>');
+    $('<a>')
+      .addClass('btn btn-secondary btn-xs ml-1 time-limit-edit-button')
+      .attr('role', 'button')
+      .attr('id', `row${row.assessment_instance_id}PopoverTimeLimit`)
+      .attr('tabindex', 0)
+      .attr('data-row', JSON.stringify(row))
+      .append($('<i class="bi-pencil-square" aria-hidden="true">'))
+      .appendTo(container);
+    value += container.html();
```

Comment:
Another instance where we could construct things with `html` (this is the last one I'll comment on).

Reply:
Updated to html template literals here and elsewhere in [306ccbe](https://github.com/PrairieLearn/PrairieLearn/pull/9725/commits/306ccbe9c0ebd4481747d7e7ddce945db6fff328) and [8475bd0](https://github.com/PrairieLearn/PrairieLearn/pull/9725/commits/8475bd0906d97cb86d2daee367e806e01396667f).